apply defaults for the config after parsing flags

rfratto opened the pull request:
Previously, ApplyDefaults was invoked immediately after the config file was loaded. This was fine prior to #107, where unmarshaling the file did not override any values specified by command line flags. However, as #107 added back UnmarshalYaml to prom.Config, the flag value was getting wiped and anything specified only by flags would fail to load.

robx commented on cmd/agent/main.go:
```diff
@@ -57,6 +57,10 @@ func main() {
 	}
 
+	if err := cfg.ApplyDefaults(); err != nil {
+		log.Fatalln("error in config file: %w", err)
```
This should probably be `log.Fatalf`.

rfratto commented:
No worries, like I said in my review I remembered that I removed UnmarshalYaml because I was running into problems with flag parsing:

> When I was working on #42 I had removed UnmarshalYaml from prom.Config because I _thought_ it was interfering with how flags were being parsed. I'm not sure what I ran into at the time, but looking at the code (and writing a few ad-hoc tests) it seems like this PR should be fine. I'll keep an eye out for any issues during testing before the next release.

This is the problem I initially had :) I wasn't running into it because the tests in `pkg/config` were unaware of flag parsing. 

> The whole setup seems kind of complicated and fragile.

I'm sorry that you're getting that impression from the code, and it's 100% my fault. Some parts of the codebase were rushed and not given a lot of care, and configuration parsing is definitely one of those, and I should've had a test for this when I removed UnmarshalYaml in the first place. This PR was rushed too, and I'll make a stronger effort to make sure each PR is done well. I'll move flag parsing to the config package and add tests that incorporate flags so we have better coverage here.

robx commented:
> I'm sorry that you're getting that impression from the code, and it's 100% my fault. Some parts of the codebase were rushed and not given a lot of care, and configuration parsing is definitely one of those, and I should've had a test for this when I removed UnmarshalYaml in the first place. This PR was rushed too, and I'll make a stronger effort to make sure each PR is done well. I'll move flag parsing to the config package and add tests that incorporate flags so we have better coverage here.

Ah don't take comment that hard, apologies. It's more that I think there's some inherent complexity in combining flags and config files and defaults which shows in the implementation and which confuses me (and I'm not sure to what extent it could be handled more clearly). It's also not an agent-specific sentiment, probably applies in much the same way to prometheus.
